packages/utils/src/conversions.ts

```diff
@@ -0,0 +1,202 @@
+    return beBigIntToHexadecimal(n)
+}
+
+/**
+ * Converts a hexadecimal string to a bigint, ensuring the hex string starts with '0x'.
+ * @param hex The hexadecimal string to convert.
+ * @returns The bigint representation of the hexadecimal string.
+ */
+export function beHexadecimalToBigint(hex: string): bigint {
+    // Ensure the hex string starts with '0x'.
+    const formattedHexString = hex.startsWith("0x") ? hex : `0x${hex}`
+
+    return BigInt(formattedHexString)
+}
+
+/**
+ * Converts a hexadecimal string to a bigint. Alias for beHexadecimalToBigint.
+ * @param hex The hexadecimal string to convert.
+ * @returns The bigint representation of the hexadecimal string.
+ */
+export function hexadecimalToBigint(hex: string): bigint {
+    return beHexadecimalToBigint(hex)
+}
+
+/**
+ * Converts a buffer to a bigint by interpreting the buffer as a hexadecimal string.
+ * @param b The buffer to convert.
+ * @returns The bigint representation of the buffer's contents.
+ */
+export function beBufferToBigint(b: Buffer): bigint {
+    return BigInt(`0x${b.toString("hex")}`)
+}
+
+/**
+ * Converts a buffer to a bigint. Alias for beBufferToBigint.
+ * @param b - The buffer to convert.
+ * @returns The bigint representation of the buffer's contents.
+ */
+export function bufferToBigint(b: Buffer): bigint {
+    return beBufferToBigint(b)
+}
+
+/**
+ * Converts a bigint to a buffer, assuming big-endian byte order, and fills with zeros if necessary.
+ * @param n The bigint to convert.
+ * @param length The number of bytes of the buffer to return.
+ * @returns The buffer representation of the bigint.
+ */
+export function beBigintToBuffer(n: bigint, length?: number): Buffer {
+    const hex = bigIntToHexadecimal(n)
+
+    // Calculate the minimum buffer length required to represent 'n' in bytes.
+    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
+    const minLength = Math.ceil(hex.length / 2)
+
+    // Use the provided length or the calculated minimum length, whichever is greater.
+    const bufferLength = length ? Math.max(length, minLength) : minLength
+
+    // Allocate buffer of the desired size, filled with zeros.
+    const buffer = Buffer.alloc(bufferLength, 0)
+
+    const fromHex = Buffer.from(hex, "hex")
+    fromHex.copy(buffer, bufferLength - fromHex.length)
+
+    return buffer
+}
+
+/**
+ * Converts a bigint to a buffer. Alias for beBigintToBuffer.
+ * @param n - The bigint to convert.
+ * @returns The buffer representation of the bigint.
+ */
+export function bigintToBuffer(n: bigint): Buffer {
+    return beBigintToBuffer(n)
+}
+
+/**
+ * Converts a BigNumberish type to a buffer, handling various input types and converting them to bigint first if necessary.
+ * @param n The BigNumberish value to convert.
+ * @returns The buffer representation of the BigNumberish value.
+ */
+export function beBigNumberishToBuffer(n: BigNumberish): Buffer {
+    if (
```

Comment:
The logic here could be delegated to `bigNumberishToBigint` to avoid duplication. That would simplify this function to only need one special case for input which is already a buffer.

  if (typeof n === "Buffer") {
    return n as Buffer;
  }
  return bigintToBuffer(bignumberishToBigint(n));

Reply:
Good catch! The following code should be fine:
```ts
 if (n instanceof Buffer) {
        return n
 }

return bigintToBuffer(bigNumberishToBigint(n))
```

I'd like to add more checks on types here. But it'll be part of another PR: https://github.com/privacy-scaling-explorations/zk-kit/pull/178 (feel free to start taking a look).